lib/Sema/DerivedConformanceKeyPathIterable.cpp

```diff
@@ -0,0 +1,136 @@
+//===--- DerivedConformanceKeyPathIterable.cpp ----------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements explicit derivation of the KeyPathIterable protocol for
+// a nominal type.
+//
+//===----------------------------------------------------------------------===//
+
+#include "CodeSynthesis.h"
+#include "TypeChecker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/GenericSignature.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/ParameterList.h"
+#include "swift/AST/Pattern.h"
+#include "swift/AST/ProtocolConformance.h"
+#include "swift/AST/Stmt.h"
+#include "swift/AST/Types.h"
+#include "DerivedConformances.h"
+
+using namespace swift;
+
+bool DerivedConformance::canDeriveKeyPathIterable(NominalTypeDecl *nominal) {
+  // There must be at least one stored property.
+  // TODO: Improve. Require struct?
```

Comment:
Yes, require struct for now. But it could also make sense for classes: each subclass implementation simply appends things to `super.allKeyPaths`.

Reply:
Changed the requirement to "must be struct" in https://github.com/apple/swift/pull/21557/commits/9b3769281937327bae90be3a24937063a0c39a47.
Also added comment about supporting classes in the future.